File: CSharp/1-TwoSum/1TwoSum.cs
// HASH MAP
// Time Complexity: O(n)
// Space Complexity: O(n) (Potentially have to add every value to the hash map)
public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        Dictionary<int, int> map = new Dictionary<int,int>();
        for (int i = 0; i < nums.Length; i++){
            int complement = target - nums[i];
            if(map.ContainsKey(complement)){
                return new int[] {map[complement], i};
            }
            map[nums[i]] = i; //add the element and it's index to the hash map. the map key is the value from nums, the map value is the index from nums.
        }
        throw new ArgumentException("No solution found.");
    }
}


//=======================================================================
// // BRUTE FORCE - TWO FOR LOOPS
// // Time Complexity: O(N^2)
// // Space Complexity O(1)
// public class Solution {
//     public int[] TwoSum(int[] nums, int target) {
//         for (int i = 0; i < nums.Length; i++){
//             for (int k = i + 1; k < nums. Length; k++){
//                 if (nums[i] + nums[k] == target){
//                     return new int[] {i, k};
//                 }
//             }
//         }
//         throw new ArgumentException("No solution found.");
//     }
// }
//=======================================================================

File: Javascript/1-TwoSum/1TwoSum.js
/* Hash Map

 // Time Complexity: O(n) where n is the number of elements in the nums array. 
 // In the worst case scenario the diff number needed is the last element so we have to traverse n-1 elements to find it. 
 // We drop constant so we get O(n). For each element we perform constant time operations (hash table lookup and insertion).

 // Space Complexity: O(n) where n is the number of elements in the nums array. 
 // In the worst case scenario the diff number needed is the last element so we store n-1 elements in the map.
 
*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    const map = new Map();
    for (let i = 0; i < nums.length; i++){
        const diff = target - nums[i];
        if (map.has(diff)) return [map.get(diff), i];
        map.set(nums[i], i);
    }
    throw new Error ("No solution for twoSum.");
};
